File: 11/seats.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const FLOOR = '.'
const EMPTY = 'L'
const TAKEN = '#'

func countVisible(seats [][]byte, i, j, visibilityRange int) int {
	res := 0
	directions := [][2]int{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, -1}, {0, 1},
		{1, -1}, {1, 0}, {1, 1},
	}
	for _, d := range directions {
		ni, nj := i+d[0], j+d[1]
		steps := 1
		for 0 <= ni && ni <= len(seats)-1 &&
			0 <= nj && nj <= len(seats[0])-1 {
			if seats[ni][nj] == EMPTY {
				break
			}
			if seats[ni][nj] == TAKEN {
				res++
				break
			}
			if visibilityRange != -1 && steps >= visibilityRange {
				break
			}
			ni += d[0]
			nj += d[1]
			steps++
		}
	}
	return res
}

func show(seats [][]byte) {
	for _, s := range seats {
		fmt.Println(string(s))
	}
	fmt.Println()
}

func nextState(seats [][]byte, new [][]byte, tolerance, visibilityLimit int) bool {
	modified := false
	rows, cols := len(seats), len(seats[0])
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			new[i][j] = seats[i][j]
			if seats[i][j] == FLOOR {
				continue
			}
			neighbours := countVisible(seats, i, j, visibilityLimit)
			if seats[i][j] == EMPTY && neighbours == 0 {
				new[i][j] = TAKEN
				modified = true
			}
			if seats[i][j] == TAKEN && neighbours >= tolerance {
				new[i][j] = EMPTY
				modified = true
			}
		}
	}
	return modified
}

// avoid memory allocations at the cost of rewriting `seats` contents.
// surprisingly does not yield any speedups
func occupiedSeatsWhenStabilized(seats [][]byte, tolerance, visibilityLimit int) int {
	modified := true
	new := make([][]byte, len(seats)) // allocate next step just once
	for i := range new {
		new[i] = make([]byte, len(seats[0]))
	}

	modified = nextState(seats, new, tolerance, visibilityLimit)
	for modified {
		seats, new = new, seats // updating the state of seats to newer one, setting up new for reuse
		// show(seats)
		modified = nextState(seats, new, tolerance, visibilityLimit)
	}

	res := 0
	for _, line := range seats {
		for _, seat := range line {
			if seat == TAKEN {
				res++
			}
		}
	}
	return res
}

func main() {
	data, _ := ioutil.ReadFile("input.txt")
	text := strings.TrimSpace(string(data))
	seats := [][]byte{}
	for _, line := range strings.Split(text, "\n") {
		seats = append(seats, []byte(line))
	}
	fmt.Println("part 1:", occupiedSeatsWhenStabilized(seats, 4, 1))

	seats = [][]byte{}
	for _, line := range strings.Split(text, "\n") {
		seats = append(seats, []byte(line))
	}
	fmt.Println("part 2:", occupiedSeatsWhenStabilized(seats, 5, -1))
}

File: 11/seats.py
from copy import deepcopy
seats = [list(s.strip()) for s in open('input.txt').readlines()]

FLOOR = '.'
EMPTY = 'L'
TAKEN = '#'


def count_taken_nearby(seats, i, j):
    res = 0
    for i_step in (-1, 0, 1):
        for j_step in (-1, 0, 1):
            if (0 <= i+i_step <= len(seats)-1 and
               0 <= j+j_step <= len(seats[0])-1 and
               not (i_step==0 and j_step==0) and
               seats[i+i_step][j+j_step] == TAKEN
               ):
                res += 1
    return res


def count_visible(seats, i, j):
    res = 0
    directions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0,  -1),          (0,  1),
        (1,  -1), (1,  0), (1 , 1)
    ]
    for (di, dj) in directions:
        ni, nj = i+di, j+dj
        while ((0 <= ni <= len(seats)-1) and
               (0 <= nj <= len(seats[0])-1)):
            if seats[ni][nj] == EMPTY:
                break
            if seats[ni][nj] == TAKEN:
                res += 1
                break
            ni, nj = ni+di, nj+dj
    return res


def next_state_1(seats):
    new = deepcopy(seats)
    for i in range(len(seats)):
        for j in range(len(seats[0])):
            if seats[i][j] == FLOOR:
                continue
            neighbours = count_taken_nearby(seats, i, j)
            if seats[i][j] == EMPTY and neighbours == 0:
                new[i][j] = TAKEN
            if seats[i][j] == TAKEN and neighbours >= 4:
                new[i][j] = EMPTY
    return new


def next_state_2(seats):
    new = deepcopy(seats)
    for i in range(len(seats)):
        for j in range(len(seats[0])):
            if seats[i][j] == FLOOR:
                continue
            neighbours = count_visible(seats, i, j)
            if seats[i][j] == EMPTY and neighbours == 0:
                new[i][j] = TAKEN
            if seats[i][j] == TAKEN and neighbours >= 5:
                new[i][j] = EMPTY
    return new


def show(seats):
    for line in seats:
        print(''.join(line))
    print()


prev_seats = deepcopy(seats)
seats = next_state_2(seats)
while seats != prev_seats:
    # show(seats)
    prev_seats = deepcopy(seats)
    seats = next_state_2(seats)

res = 0
for line in seats:
    for seat in line:
        if seat == TAKEN:
            res += 1
print(res)
